### ISSUE TITLE
[datasets] `random_shuffle` overspills objects on random node

### ISSUE THREAD

**Issue by krfricke** [ray-project/ray#17612]
### What is the problem?

When using `random_shuffle`, the object store spills objects even though the disk is full. This leads to filesystem errors.

Usually, objects are only spilled on one of the nodes. In the cluster setup, the head node has 200 GB storage space and the worker nodes 100 GB. One worker node seems to consistently overspill their disk.

### Reproduction (REQUIRED)

`simple_spill.py`
```
import time
import numpy as np
import torch
import ray
import argparse
import os

DUMMY_ROW = {'key': 0, 'embeddings_name0': 1527, 'embeddings_name1': 175, 'embeddings_name2': 8, 'embeddings_name3': 5, 'embeddings_name4': 5, 'embeddings_name5': 687, 'embeddings_name6': 165, 'embeddings_name7': 10, 'embeddings_name8': 137, 'embeddings_name9': 597, 'embeddings_name10': 1574, 'embeddings_name11': 78522, 'embeddings_name12': 283941, 'embeddings_name13': 3171, 'embeddings_name14': 39560, 'embeddings_name15': 718571, 'embeddings_name16': 73699, 'one_hot0': 1, 'one_hot1': 30, 'labels': 0.3801173847541949, '__index_level_0__': 0}


if __name__ == '__main__':
    ray.init(address="auto")

    num_workers = 4

    pipeline = ray.data.range(200_000_000).map(lambda i: DUMMY_ROW)
    pipeline = pipeline.repeat().random_shuffle()  # Works if `random_shuffle` is not used
    splits = pipeline.split(num_workers, equal=True)

    @ray.remote
    def consume(split, rank=None, batch_size=None):
        for i, x in enumerate(split.iter_rows()):
            time.sleep(1)
            if i % 10 == 0:
                print(i)
        return

    tasks = [
        consume.remote(split, rank=0, batch_size=250000)
        for idx, split in enumerate(splits)
    ]
    ray.get(tasks)
```

`spill_cluster.yaml`
```
cluster_name: shuffling-data-loader-horovod-3

min_workers: 3
initial_workers: 3
max_workers: 3

idle_timeout_minutes: 20
docker:
    image: horovod/horovod-ray:sha-3278964 # 7/20
    container_name: ray_docker

provider:
    type: aws
    region: us-west-2
    cache_stopped_nodes: False

auth:
    ssh_user: ubuntu

head_node:
    InstanceType: i3.xlarge
    ImageId: latest_dlami
    KeyName: kai-anyscale
    BlockDeviceMappings:
        - DeviceName: /dev/sda1
          Ebs:
              VolumeSize: 200
    TagSpecifications:
        - ResourceType: "instance"
          Tags:
              - Key: anyscale-user
                Value: "kai@anyscale.com"


worker_nodes:
    InstanceType: i3.xlarge
    ImageId: latest_dlami
    KeyName: richard-anyscale
    IamInstanceProfile:
        Arn: arn:aws:iam::959243851260:instance-profile/ray-autoscaler-v1
    TagSpecifications:
        - ResourceType: "instance"
          Tags:
              - Key: anyscale-user
                Value: "kai@anyscale.com"

file_mounts: {
    "/test": ".",
}

setup_commands:
    - pip install -q boto3 fsspec==2021.05.0 s3fs==2021.05.0 tqdm torch torchvision tensorboard pyarrow
    - pip uninstall ray -y && pip install -U https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-2.0.0.dev0-cp37-cp37m-manylinux2014_x86_64.whl

head_setup_commands: []

worker_setup_commands: []

head_start_ray_commands:
    - ray stop --force
    -  ray start --head --port=6379 --object-manager-port=8076 --autoscaling-config=~/ray_bootstrap_config.yaml --object-store-memory=$(( 8 * 1000 * 1000 * 1000 ))

worker_start_ray_commands:
    - ray stop --force
    - RAY_BACKEND_LOG_LEVEL=debug ray start --address=$RAY_HEAD_IP:6379 --object-manager-port=8076 --object-store-memory=$(( 8 * 1000 * 1000 * 1000 ))
```

cc @richardliaw @ericl 

- [x] I have verified my script runs in a clean environment and reproduces the issue.
- [x] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

**Comment by krfricke** [comment]
Note that in my tests this usually only happened on a single worker node (not all of them) - only one node would try to spill objects to disk. 

Also, if the `random_shuffle()` call is commented out, I never saw this issue, so it might be related to that.

**Comment by rkooo567** [comment]
So, if we raise the exception and crash the workload, is it the expected behavior?

**Comment by richardliaw** [comment]
Actually, not really. What's the operating rule of thumb here (for disk space vs data size)?

**Comment by ericl** [comment]
Two issues:
- How do we raise an error message here
- How do we avoid this imbalance scenario
It would be good to review how other systems handle disk space full (iirc, stop accepting tasks is a common solution but may not work for us)

**Comment by rkooo567** [comment]
For this particular issue, the disk usage was imbalanced because all consumers are scheduled on the same node and ingesting there. Since these tasks are long-running (while loop), all of shuffled data was coming to that node and pinned, which causes high disk usage.

We definitely need to handle high disk usage in a better way. So I will keep the issue open for tracking purpose.

**Comment by zhe-thoughts** [comment]
(Context: I’m doing a round of cleanup of waiting-for-triage issues)

Marking this as P1 and assigning to myself (it's already in Core Backlog). And let's prioritize it according to global list. 

cc @rkooo567 @scv119 @stephanie-wang to correct me on the priority if needed

**Comment by amogkam** [comment]
Doing cleanup of data issues and going to move this back to Core as this looks like a core issue.

**Comment by cszhu** [comment]
Hello! This P2 issue has seen no activity in at least 2 years. It will be closed in 2 weeks as part of ongoing cleanup efforts.

Please remove the `pending-cleanup` label if you believe this issue should remain open.

Thanks for contributing to Ray!
